apps/meteor/server/lib/findUsersOfRoomOrderedByRole.ts

```diff
@@ -0,0 +1,188 @@
+		...(sort || defaultSort),
+	};
+
+	const userLookupPipeline: Document[] = [{ $match: { $expr: { $eq: ['$_id', '$$userId'] } } }];
+
+	if (status) {
+		userLookupPipeline.push({ $match: { status } });
+	}
+
+	userLookupPipeline.push({
+		$match: {
+			$and: [
+				{
+					active: true,
+					username: {
+						$exists: true,
+						...(exceptions.length > 0 && { $nin: exceptions }),
+					},
+					...(filter && orStmt.length > 0 && { $or: orStmt }),
+				},
+				...extraQuery,
+			],
+		},
+	});
+
+	userLookupPipeline.push({
+		$project: {
+			_id: 1,
+			username: 1,
+			name: 1,
+			nickname: 1,
+			status: 1,
+			avatarETag: 1,
+			_updatedAt: 1,
+			federated: 1,
+			statusConnection: 1,
+		},
+	});
+
+	const defaultPriority = rolesInOrder.length + 1;
+
+	const branches = rolesInOrder.map((role, index) => ({
+		case: { $eq: ['$$this', role] },
+		then: index + 1,
+	}));
+
+	const filteredPipeline: Document[] = [
+		{
+			$lookup: {
+				from: 'users',
+				let: { userId: '$u._id' },
+				pipeline: userLookupPipeline,
+				as: 'userDetails',
+			},
+		},
+		{ $unwind: '$userDetails' },
+		{
+			$addFields: {
+				primaryRole: {
+					$reduce: {
+						input: '$roles',
+						initialValue: { role: null, priority: defaultPriority },
+						in: {
+							$let: {
+								vars: {
+									currentPriority: {
+										$switch: {
+											branches,
+											default: defaultPriority,
+										},
+									},
+								},
+								in: {
+									$cond: [
+										{
+											$and: [{ $in: ['$$this', rolesInOrder] }, { $lt: ['$$currentPriority', '$$value.priority'] }],
+										},
+										{ role: '$$this', priority: '$$currentPriority' },
+										'$$value',
+									],
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		{
+			$addFields: {
+				rolePriority: { $ifNull: ['$primaryRole.priority', defaultPriority] },
+			},
+		},
+		{
+			$project: {
+				_id: '$userDetails._id',
+				rid: 1,
+				roles: 1,
+				primaryRole: '$primaryRole.role',
+				rolePriority: 1,
+				username: '$userDetails.username',
+				name: '$userDetails.name',
+				nickname: '$userDetails.nickname',
+				status: '$userDetails.status',
+				avatarETag: '$userDetails.avatarETag',
+				_updatedAt: '$userDetails._updatedAt',
+				federated: '$userDetails.federated',
+				statusConnection: '$userDetails.statusConnection',
+			},
+		},
+	];
+
+	const facetPipeline: Document[] = [
+		{ $match: { rid } },
+		{
+			$facet: {
+				totalCount: [{ $match: { rid } }, ...filteredPipeline, { $count: 'total' }],
+				members: [
+					{ $match: { rid } },
+					...filteredPipeline,
+					{ $sort: sortCriteria },
+					...(skip > 0 ? [{ $skip: skip }] : []),
+					...(limit > 0 ? [{ $limit: limit }] : []),
+				],
+			},
+		},
+		{
+			$project: {
+				members: 1,
+				totalCount: { $arrayElemAt: ['$totalCount.total', 0] },
+			},
+		},
+	];
+
+	const [result] = await Subscriptions.col.aggregate(facetPipeline, { allowDiskUse: true }).toArray();
```

Comment:
@MarcosSpessatto , should we run this on the secondary like the other aggregations?

Reply:
I think it makes sense